SnoopCompileCore/src/snoop_inference.jl

```diff
@@ -1,100 +1,26 @@
 export @snoop_inference
 
-struct InferenceTiming
-    mi_info::Core.Compiler.Timings.InferenceFrameInfo
-    inclusive_time::Float64
-    exclusive_time::Float64
-end
-"""
-    inclusive(frame)
-
-Return the time spent inferring `frame` and its callees.
-"""
-inclusive(it::InferenceTiming) = it.inclusive_time
-"""
-    exclusive(frame)
-
-Return the time spent inferring `frame`, not including the time needed for any of its callees.
-"""
-exclusive(it::InferenceTiming) = it.exclusive_time
-
-struct InferenceTimingNode
-    mi_timing::InferenceTiming
-    start_time::Float64
-    children::Vector{InferenceTimingNode}
-    bt
-    parent::InferenceTimingNode
-
-    # Root constructor
-    InferenceTimingNode(mi_timing::InferenceTiming, start_time, @nospecialize(bt)) =
-        new(mi_timing, start_time, InferenceTimingNode[], bt)
-    # Child constructor
-    function InferenceTimingNode(mi_timing::InferenceTiming, start_time, @nospecialize(bt), parent::InferenceTimingNode)
-        child = new(mi_timing, start_time, InferenceTimingNode[], bt, parent)
-        push!(parent.children, child)
-        return child
-    end
-end
-inclusive(node::InferenceTimingNode) = inclusive(node.mi_timing)
-exclusive(node::InferenceTimingNode) = exclusive(node.mi_timing)
-InferenceTiming(node::InferenceTimingNode) = node.mi_timing
-
-function InferenceTimingNode(t::Core.Compiler.Timings.Timing)
-    ttree = timingtree(t)
-    it, start_time, ttree_children = ttree::Tuple{InferenceTiming, Float64, Vector{Any}}
-    root = InferenceTimingNode(it, start_time, t.bt)
-    addchildren!(root, t, ttree_children)
-    return root
+const snoop_inference_lock = ReentrantLock()
+const newly_inferred = CodeInstance[]
+
+function start_tracking()
+    islocked(snoop_inference_lock) && error("already tracking inference (cannot nest `@snoop_inference` blocks)")
```

Comment:
(this really should be a Base feature, since implementing it indirectly in packages is not very robust)
```suggestion
    snoop_inference_lock.reentrancy_cnt == 1 || throw(ConcurrencyViolationError("already tracking inference (cannot nest `@snoop_inference` blocks)"))
```